src/PowerShellEditorServices/Services/DebugAdapter/DebugService.cs

```diff
@@ -792,55 +805,95 @@ private bool AddToAutoVariables(PSObject psvariable, string scope)
         private async Task FetchStackFramesAsync(string scriptNameOverride)
         {
             PSCommand psCommand = new PSCommand();
+            // The serialization depth to retrieve variables from remote runspaces.
+            var serializationDepth = 3;
 
             // This glorious hack ensures that Get-PSCallStack returns a list of CallStackFrame
             // objects (or "deserialized" CallStackFrames) when attached to a runspace in another
             // process.  Without the intermediate variable Get-PSCallStack inexplicably returns
             // an array of strings containing the formatted output of the CallStackFrame list.
-            var callStackVarName = $"$global:{PsesGlobalVariableNamePrefix}CallStack";
-            psCommand.AddScript($"{callStackVarName} = Get-PSCallStack; {callStackVarName}");
+            var callStackVarName = $"$GLOBAL:{PsesGlobalVariableNamePrefix}CallStack";
 
-            var results = await _executionService.ExecutePSCommandAsync<PSObject>(psCommand, CancellationToken.None).ConfigureAwait(false);
+            var getPSCallStack = $"Get-PSCallStack | % {{ [void]{callStackVarName}.add(@($PSItem,$PSItem.GetFrameVariables())) }}";
+
+            // If we're attached to a remote runspace, we need to serialize the callstack prior to transport
+            // because the default depth is too shallow
+            var isOnRemoteMachine = _psesHost.CurrentRunspace.IsOnRemoteMachine;
+            var returnSerializedIfOnRemoteMachine = isOnRemoteMachine
+                ? $"[Management.Automation.PSSerializer]::Serialize({callStackVarName}, {serializationDepth})"
+                : callStackVarName;
+
+            // We have to deal with a shallow serialization depth with ExecutePSCommandAsync as well, hence the serializer to get full var information
+            // TODO: Don't use serializer for local runspaces, or implement a way to specify depth ExecutePSCommandAsync
+            string getPSCallStackScript = $"[Collections.ArrayList]{callStackVarName}=@();{getPSCallStack};{returnSerializedIfOnRemoteMachine}";
+            psCommand.AddScript(getPSCallStackScript);
 
-            var callStackFrames = results.ToArray();
 
-            this.stackFrameDetails = new StackFrameDetails[callStackFrames.Length];
+            // PSObject is used here instead of the specific type because we get deserialized objects from remote sessions and want a common interface
+            var results = await _executionService.ExecutePSCommandAsync<PSObject>(psCommand, CancellationToken.None).ConfigureAwait(false);
+
+            IEnumerable callStack = isOnRemoteMachine
+                ? (PSSerializer.Deserialize(results[0].BaseObject as string) as PSObject).BaseObject as IList
```

Comment:
@SeeminglyScience where should this utility method be located? In this file or a separate utility file?

Reply:
@SeeminglyScience Would this be a blocker or should we get this merged and then make that change in another (smaller) PR?

Reply:
Sorry was already out and about! Yeah it didn't have to be a blocker, my bad.  Anyway I'd either put it as an extension method in `src\PowerShellEditorServices\Utility\Extensions.cs` called `TryUnwrapPSObjectAs` or create a new file in that folder called `PSObjectUtils.cs` and make it a static method. I sort of lean towards the latter just because extension methods on `object` add to intellisense for *everything* and there's no way to say "only when strictly typed as `object`".

Reply:
I agree with the TryUnwrap utility method, although it could use `switch` instead of `if`.  If it is only currently used here then I feel it can just be a private method in a `Utility` region.